wasmtime-runtime/src/jit_dump.rs

```diff
@@ -0,0 +1,701 @@
+    count: u64,
+    // DebugEntry array is appeneded during writing.
+}
+
+/*
+ *  Each jitdump file starts with a fixed size header containing the following fields in order:
+ *
+ * uint32_t magic     : a magic number tagging the file type. The value is 4-byte long and represents the string "JiTD" in ASCII form. It is 0x4A695444 or 0x4454694a depending on the endianness. The field can be used to detect the endianness of the file
+ * uint32_t version   : a 4-byte value representing the format version. It is currently set to 2
+ * uint32_t total_size: size in bytes of file header
+ * uint32_t elf_mach  : ELF architecture encoding (ELF e_machine value as specified in /usr/include/elf.h)
+ * uint32_t pad1      : padding. Reserved for future use
+ * uint32_t pid       : JIT runtime process identification (OS specific)
+ * uint64_t timestamp : timestamp of when the file was created
+ * uint64_t flags     : a bitmask of flags
+ */
+
+#[derive(Serialize, Deserialize, Debug, Default, IOwrite, SizeWith)]
+#[repr(C)]
+pub struct FileHeader {
+    magic: u32,
+    version: u32,
+    size: u32,
+    e_machine: u32,
+    pad1: u32,
+    pid: u32,
+    timestamp: u64,
+    flags: u64,
+}
+
+extern "C" {
+    fn clock_gettime(clk_id: clockid_t, tp: *mut timespec) -> c_int;
+}
+
+/// Interface for driving the creation of jitdump files
+#[derive(Debug)]
+pub struct JitDumpAgent {
+    jitdump_file: File,
+    code_index: u64,
+    dump_funcs: bool,
+}
+
+impl JitDumpAgent {
+    /// Intialize a JitDumpAgent and write out the header
+    pub fn new() -> Result<Self, Error> {
+        let filename = format!("./jit-{}.dump", process::id());
+        let mut jitdump_file;
+        unsafe {
+            let filename_c = CString::new(filename)?;
+            let fd = open(
+                filename_c.as_ptr(),
+                libc::O_CREAT | libc::O_TRUNC | libc::O_RDWR,
+                0666,
+            );
+            let pgsz = sysconf(libc::_SC_PAGESIZE) as usize;
+            mmap(
+                0 as *mut c_void,
+                pgsz,
+                libc::PROT_EXEC | libc::PROT_READ,
+                libc::MAP_PRIVATE,
+                fd,
+                0,
+            );
+            jitdump_file = File::from_raw_fd(fd);
+        }
+        JitDumpAgent::write_file_header(&mut jitdump_file)?;
+
+        Ok(Self {
+            jitdump_file: jitdump_file,
+            code_index: 0,
+            dump_funcs: true,
+        })
+    }
+
+    fn get_time_stamp(timestamp: &mut u64) -> c_int {
+        unsafe {
+            let mut ts = mem::MaybeUninit::zeroed().assume_init();
+            if clock_gettime(libc::CLOCK_MONOTONIC, &mut ts) != 0 {
+                println!("Error getting timestamp");
+            }
+            // TODO: What does it mean for either sec or nsec to be negative?
+            *timestamp = (ts.tv_sec * 1000000000 + ts.tv_nsec) as u64;
+        }
+        return 0;
+    }
+
+    // Get the ELF machine architecture.
```

Comment:
Please use triple slash. When telling rustdoc to document private items, it is useful